tests/conftest.py

```diff
@@ -21,53 +29,92 @@ def create_token(deployer):
-def validators(create_validator, request):
-    return [create_validator() for _ in range(int(request.param.split(" ")[0]) + 1)]
+@pytest.fixture(scope="session")
+def validator(create_validator):
+    return create_validator()
 
 
 @pytest.fixture(scope="session")
 def MIN_STREAM_LIFE():
-    return 60 * 60  # 1 hour in seconds
+    return ONE_HOUR
 
 
 @pytest.fixture(scope="session")
-def stream_manager_contract(owner, project, MIN_STREAM_LIFE, validators, tokens):
-    return owner.deploy(project.StreamManager, owner, MIN_STREAM_LIFE, validators, tokens)
+def stream_manager_contract(project, controller, token, validator, MIN_STREAM_LIFE):
+    return project.StreamManager.deploy(
+        controller,
+        int(MIN_STREAM_LIFE.total_seconds()),
+        [token],
+        [validator],
+        sender=controller,
+    )
 
 
 @pytest.fixture(scope="session")
 def stream_manager(stream_manager_contract):
     return StreamManager(stream_manager_contract)
 
 
+@pytest.fixture(scope="session", params=["1 product", "2 products", "3 products"])
+def products(request):
+    return [
+        # NOTE: 0x[25 empty bytes]01 ~= 0.00028... tokens/second ~= 1.01... tokens/hr
+        #       also, `sum(1, 2, 3, ..., n) = n * (n - 1) / 2`
+        HashBytes32(b"\x00" * 25 + to_bytes(product_code) + b"\x00" * 6)
+        for product_code in range(1, int(request.param.split(" ")[0]) + 1)
+    ]
+
+
+@pytest.fixture(scope="session", params=["1 hour", "2 hours", "12 hours"])
+def stream_life(request):
+    return int(request.param.split(" ")[0]) * ONE_HOUR
+
+
+@pytest.fixture(scope="session")
+def funding_rate(token, products):
+    return Decimal(sum(map(to_int, products))) / Decimal(10 ** token.decimals())
+
+
+@pytest.fixture(scope="session")
+def create_stream(chain, stream_manager, token, payer, products, stream_life, funding_rate):
+    def create_stream(
+        amount: int | None = None,
+        sender: AccountAPI | None = None,
+        allowance: int = (2**256 - 1),
+        **txn_args,
+    ):
+        if amount is None:
+            amount = int(
+                Decimal(stream_life.total_seconds())
+                * funding_rate
+                # NOTE: To undo the adjustment factor from above
+                * Decimal(10 ** token.decimals())
+            )
+            assert amount <= token.balanceOf(sender or payer)
+
+        if token.allowance(sender or payer, stream_manager.address) != allowance:
+            token.approve(stream_manager.address, allowance, sender=(sender or payer))
+
+        return stream_manager.create(token, amount, products, sender=(sender or payer), **txn_args)
+
+    return create_stream
+
+
 @pytest.fixture(scope="session")
-def stream(stream_manager, token, payer):
-    return stream_manager.create(token, 1000, sender=payer)
+def stream(chain, create_stream):
+    # TODO: Remove when https://github.com/ApeWorX/ape/pull/2277 merges
```

Comment:
sorry i see you made some changes here since i had figured this out.
but moving the isolation to session scoped stream fixture doesnt change much.

the interlocking of those parametrized tests causes things to run in that order.
create stream will not be 0 becase cancel stream would have triggered the a session scoped fixure

but making the session scoped fixture wrapped in isolation, it essentially is undoing itself much like a function scoped fixture

Reply:
yes, I think this is a good point, that this fixture should not be `session`